packages/webapp/src/store/api/apiSlice.ts

```diff
@@ -0,0 +1,55 @@
+/*
+ *  Copyright 2024 LiteFarm.org
+ *  This file is part of LiteFarm.
+ *
+ *  LiteFarm is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  LiteFarm is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
+ */
+
+import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
+import { RootState } from '../store';
+import { animalsUrl, animalBatchesUrl, animalGroupsUrl, url } from '../../apiConfig';
+import type { Animal, AnimalBatch, AnimalGroup } from './types';
+
+export const api = createApi({
+  baseQuery: fetchBaseQuery({
+    baseUrl: url,
+    prepareHeaders: (headers, { getState }) => {
+      const state = getState() as RootState;
+
+      headers.set('Content-Type', 'application/json');
+      headers.set('Authorization', `Bearer ${localStorage.getItem('id_token')}`);
+      headers.set('user_id', state.entitiesReducer.userFarmReducer.user_id || '');
+      headers.set('farm_id', state.entitiesReducer.userFarmReducer.farm_id || '');
+
+      return headers;
+    },
+    responseHandler: 'content-type',
+  }),
+  tagTypes: ['Animals', 'AnimalBatches', 'AnimalGroups'],
+  endpoints: (build) => ({
+    // redux-toolkit.js.org/rtk-query/usage-with-typescript#typing-query-and-mutation-endpoints
+    // <ResultType, QueryArg>
+    getAnimals: build.query<Animal[], void>({
+      query: () => `${animalsUrl}`,
+      providesTags: ['Animals'],
```

Comment:
Just wondering are we waiting until we have more endpoints to do specific animal tags? 

To get both individual animal tags and we could do : 
```
(result, error, userFarm) => {
  const tags= result.map((animal) => return {type: 'Animal', id: animal.id});
  tags.push {return type: 'FarmAnimals', id: userFarm.farm_id}
}
```

Just have to pass in useFarm on the query selector.

Reply:
Oh shoot I meant to ask about this in meeting this morning because I don't quite follow.

The idea is that you can create a new tag and entity (the singular animal) based upon the `/animals` endpoint, without ever having to have an actual `GET` singular animal endpoint? I didn't know that was a pattern, but it sounds interesting.

What I don't follow is the role of `userFarm` though, since the base endpoint was already only for one farm. I would think the nested tag would want to be associated with `animal.id`, not `farm_id`.

Reply:
The idea is you can invalidate tags for an individual animal (update, remove). The backend endpoint/query param will still be needed. 

For 'FarmAnimals' tag -- its essentially the same as ['Animals'] you have currently as its goal would be to refresh all animals.  I just chose a structure which is probably not necessary.... The example I have from my class uses users.id to go with something like 'UsersMovies' tag. I was just transposing.